### PR TITLE
GH-2412: Improved JoinClassifier for unbound values

### LINKED ISSUE #2412: Incorrect JoinClassifier results with unbound values.
### Version

5.1.0-SNAPSHOT

### What happened?

The following two variants incorrectly return 1 binding whereas 0 would be expected.
The reason is that JoinClassifier.isLinear returns true because VarFinder does not consider the unbound values.

A linear join substitutes the (in-scope) variables on the right-hand-side (rhs) with each binding from the left-hand-side (lhs). This causes the incorrect substitution of ?x on the rhs with `0` such that `bound(?x)` yields true.

```sparql
SELECT * {
  VALUES ?x {
    0
  }
  {
    VALUES ?x {
      UNDEF
    }
    FILTER(bound(?x))
  }
}
```

```sparql
SELECT * {
  BIND(0 AS ?x)
  {
    BIND(coalesce() AS ?x)
    FILTER(bound(?x))
  }
}
```

### Relevant output and stacktrace

_No response_

### Are you interested in making a pull request?

Yes

### PR DESCRIPTION
**Pull request by Aklakan**
GitHub issue resolved #2412

Pull request Description: Additional heuristics and test cases in JoinClassifier and VarFinder for dealing with unbound values.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java [hunk 1]
```diff
@@ -351,9 +389,26 @@ public void visit(OpProject opProject) {
...
+                while (varIt.hasNext()) {
+                    Var var = varIt.next();
+                    if (!row.contains(var)) {
+                        varIt.remove();
+                        optVars.add(var);
+                    }
+                }
+            }
+            defines.addAll(definedVars);
+            optDefines.addAll(optVars);
```
Comment (Aklakan) [inline jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:411]: Are the arguments against treating unbound table values as optDefines?

File: jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java [hunk 2]
```diff
@@ -204,7 +204,17 @@ public void dynamic_dft_7() {
                   "{ GRAPH <urn:x-arq:DefaultGraph> { ?s <uri:p> ?o . FILTER ( ?o IN ( 1, 2) ) } }",
                   2, dataset);
     }
-    
+
+    @Test
```
Comment (Aklakan) [inline jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java:208]: This class is not an ideal fit for the join tests. Any suggestion?

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java [hunk 3]
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
 
         private void processAssignVarExprList(VarExprList varExprList) {
             varExprList.forEachVarExpr((v,e)-> {
-                defines.add(v) ; // Expression may eval to error -> unset?
-                if ( e != null )
+                if ( e != null ) {
+                    boolean isLikelyToBeDefined = isLikelyToBeDefined(e);
+                    if (isLikelyToBeDefined) {
+                        defines.add(v) ; // Expression may eval to error -> unset?
```
Comment (Aklakan) [inline jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:342]: If the variable may be unbound it probably still needs to be tracked - e.g. in optDefines?

**Comment by Aklakan** [comment]
I just removed some left-over code that had slipped into this PR.
The proposed changes are ready for discussion.

File: jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java (hunk 2 above)
Comment (rvesse) [reply jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java:208]: Can you not make the `testCount()` method `public static` and then create a new test class that simply calls it?

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java [hunk 4]
```diff
@@ -351,9 +389,26 @@ public void visit(OpProject opProject) {
...
+            Table table = opTable.getTable();
+            Set<Var> definedVars = new HashSet<>(table.getVars());
+            List<Var> optVars = new ArrayList<>();
+            Iterator<Binding> rowIt = table.rows();
+            while (rowIt.hasNext()) {
+                Binding row = rowIt.next();
+                Iterator<Var> varIt = definedVars.iterator();
+                while (varIt.hasNext()) {
+                    Var var = varIt.next();
+                    if (!row.contains(var)) {
```
Comment (rvesse) [inline jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:404]: This seems overly aggressive at first glance.  If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked. What happens if you have multiple rows with some values and some `UNDEF`, is that going to create bad results if we allow linearisation?

Perhaps a different solution is needed around handling of linear joins for `OpTable` to just never substitute over an `UNDEF` though guess that might be significantly more involved?

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java (hunk 1 above)
Comment (rvesse) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:411]: No, makes sense to me

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java (hunk 4 above)
Comment (Aklakan) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:404]: > If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked.

Its per variable - so if there is an UNDEF value for a variable then linearization for that variable gets blocked.
Conversely, if the variable is defined in all rows its added to the defined set.
I was creating test cases for skewed multi-variable joins (i.e. joins on multiple variables where some of them are unbound) when I noticed that I get incorrect results.

File: jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java (hunk 2 above)
Comment (Aklakan) [reply jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java:208]: With my question I was mainly referring to whether I have overlooked an already existing suitable place for these tests - if there is none so far I could of course create a new class and register it with the test suite.

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java (hunk 3 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:342]: Yes - that seems likely. 
Not putting it in is quite safe.
Doing so and tests all pass.

Comment (Aklakan) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:343]: Done

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java (hunk 1 above)
Comment (Aklakan) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java:423]: Added variable to optDefines if it is not considered defined.

File: jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java (hunk 2 above)
Comment (Aklakan) [reply jena-arq/src/test/java/org/apache/jena/sparql/core/AbstractTestDynamicDataset.java:208]: There is now a AbstractTestQueryExec hierarchy analogous to AbstractTestDynamicDataset.

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java [hunk 5]
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
...
+            boolean result;
+            if (expr.isVariable()) {
+                result = defines.contains(expr.asVar());
+            } else if (expr.isConstant() || expr instanceof E_Bound) {
+                result = true;
+            } else if (expr.isFunction()) {
+                ExprFunction fn = expr.getFunction();
+                List<Expr> args = fn.getArgs();
+                if (fn instanceof E_Coalesce) {
+                    result = args.stream().anyMatch(this::isLikelyToBeDefined);
```
[the remark this answers is not in the document]
Comment (Aklakan) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java]: Done

File: jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java [hunk 6]
```diff
@@ -329,12 +336,43 @@ public void visit(OpExtend opExtend) {
...
+                } else {
+                    result = args.stream().allMatch(this::isLikelyToBeDefined);
```
[the remark this answers is not in the document]
Comment (Aklakan) [reply jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java]: Done

**Comment by Aklakan** [comment]
I think I addressed all changed. You probably still want to check the AbstractTestQueryExec / TestQueryExec* classes that I added.

**Comment by afs** [comment]
Looks good.

Attached are some more tests `TestVarFinder2` for `VarFinder` directly.
If you agree with my understanding, please add the file and add to `TS_Algebra`.
(5 tests fail in Jena without this PR).
[TestVarFinder2.java.txt](https://github.com/apache/jena/files/15049216/TestVarFinder2.java.txt)

One clearing thing:
`AbstractTestDynamicDataset` - the `join_*` test are still there.

**Comment by Aklakan** [comment]
The test cases in TestVarFinder2 look good to me.
I added one more assertion to `varfind_extend_6`: `assertSetVars(vf.getOpt(), var_x);`

> AbstractTestDynamicDataset - the join_* test are still there.

Whoops, removed.

**Review by afs — APPROVED** [review]

**Comment by afs** [comment]
Thanks!

**Pull request merged**